scan/base.go

```diff
@@ -40,6 +44,378 @@ type base struct {
+		if err = json.Unmarshal([]byte(string(body)), &jsonError); err != nil {
+			return
+		}
+		if jsonError.Error == "HTTP Token: Access denied.\n" {
+			return vinfos, fmt.Errorf("wordpress: HTTP Token: Access denied.")
+		} else {
+			return vinfos, fmt.Errorf("status: %s", resp.Status)
+		}
+	}
+	coreConvertVinfos(string(body))
+	return
+}
+
+func coreConvertVinfos(stdout string) (vinfos []models.VulnInfo, err error) {
+	data := map[string]WpCveInfos{}
+	if err = json.Unmarshal([]byte(stdout), &data); err != nil {
+		var jsonError WpCveInfos
+		if err = json.Unmarshal([]byte(stdout), &jsonError); err != nil {
+			return
+		}
+	}
+	for _, e := range data {
+		if len(e.Vulnerabilities) == 0 {
+			continue
+		}
+		for _, vulnerability := range e.Vulnerabilities {
+			if len(vulnerability.References.Cve) == 0 {
+				continue
+			}
+			notFixedYet := false
+			if len(vulnerability.FixedIn) == 0 {
+				notFixedYet = true
+			}
+			var cveIDs []string
+			for _, cveNumber := range vulnerability.References.Cve {
+				cveIDs = append(cveIDs, "CVE-"+cveNumber)
+			}
+
+			for _, cveID := range cveIDs {
+				vinfos = append(vinfos, models.VulnInfo{
+					CveID: cveID,
+					CveContents: models.NewCveContents(
+						models.CveContent{
+							CveID: cveID,
+							Title: vulnerability.Title,
+						},
+					),
+					AffectedPackages: models.PackageStatuses{
+						{
+							NotFixedYet: notFixedYet,
+						},
+					},
+				})
+			}
+		}
+	}
+	return
+}
+
+//WpStatus is for wp command
+type WpStatus struct {
+	Name    string `json:"name"`
+	Status  string `json:"status"`
+	Update  string `json:"update"`
+	Version string `json:"version"`
+}
+
+func detectWpTheme(c *base) (vinfos []models.VulnInfo, err error) {
+	cmd := fmt.Sprintf("wp theme list --path=%s --format=json", c.ServerInfo.WpPath)
+
+	var themes []WpStatus
+	var r execResult
+	if r = exec(c.ServerInfo, cmd, noSudo); r.isSuccess() {
+		if err = json.Unmarshal([]byte(r.Stdout), &themes); err != nil {
+			return
+		}
+	}
+	if !r.isSuccess() {
+		return vinfos, fmt.Errorf("%s", cmd)
+	}
+
+	for _, theme := range themes {
+		url := fmt.Sprintf("https://wpvulndb.com/api/v3/themes/%s", theme.Name)
+		token := fmt.Sprintf("Token token=%s", c.ServerInfo.WpToken)
+		var req *http.Request
+		req, err = http.NewRequest("GET", url, nil)
+		if err != nil {
+			return
+		}
+		req.Header.Set("Authorization", token)
+		client := new(http.Client)
+		var resp *http.Response
+		resp, err = client.Do(req)
+		if err != nil {
+			return
+		}
+		body, _ := ioutil.ReadAll(resp.Body)
+		defer resp.Body.Close()
+		if resp.StatusCode != 200 && resp.StatusCode != 404 {
+			return vinfos, fmt.Errorf("status: %s", resp.Status)
+		} else if resp.StatusCode == 404 {
+			var jsonError WpCveInfos
+			if err = json.Unmarshal([]byte(string(body)), &jsonError); err != nil {
+				return
+			}
+			if jsonError.Error == "HTTP Token: Access denied.\n" {
+				return vinfos, fmt.Errorf("wordpress: HTTP Token: Access denied.")
```

Comment:
__[Golint]__ error strings should not be capitalized or end with punctuation or a newline

<sub>posted by Sider</sub>